rust/datafusion/src/datasource/parquet.rs

```diff
@@ -0,0 +1,586 @@
+                                }
+                            }
+                            Arc::new(builder.finish())
+                        }
+                        ColumnReader::FloatColumnReader(ref mut r) => {
+                            ArrowReader::<Float32Type>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::DoubleColumnReader(ref mut r) => {
+                            ArrowReader::<Float64Type>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::FixedLenByteArrayColumnReader(ref mut r) => {
+                            read_binary_column!(self, r, i)
+                        }
+                        ColumnReader::ByteArrayColumnReader(ref mut r) => {
+                            read_binary_column!(self, r, i)
+                        }
+                    };
+
+                    batch.push(array);
+                }
+
+                if batch.len() == 0 || batch[0].data().len() == 0 {
+                    Ok(None)
+                } else {
+                    Ok(Some(RecordBatch::try_new(
+                        self.projection_schema.clone(),
+                        batch,
+                    )?))
+                }
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+/// Create a new schema by applying a projection to this schema's fields
+fn schema_projection(schema: &Schema, projection: &[usize]) -> Result<Arc<Schema>> {
+    let mut fields: Vec<Field> = Vec::with_capacity(projection.len());
+    for i in projection {
+        if *i < schema.fields().len() {
+            fields.push(schema.field(*i).clone());
+        } else {
+            return Err(ExecutionError::InvalidColumn(format!(
+                "Invalid column index {} in projection",
+                i
+            )));
+        }
+    }
+    Ok(Arc::new(Schema::new(fields)))
+}
+
+/// convert a Parquet INT96 to an Arrow timestamp in nanoseconds
+fn convert_int96_timestamp(v: &[u32]) -> i64 {
+    const JULIAN_DAY_OF_EPOCH: i64 = 2_440_588;
+    const SECONDS_PER_DAY: i64 = 86_400;
+    const MILLIS_PER_SECOND: i64 = 1_000;
+
+    let day = v[2] as i64;
+    let nanoseconds = ((v[1] as i64) << 32) + v[0] as i64;
+    let seconds = (day - JULIAN_DAY_OF_EPOCH) * SECONDS_PER_DAY;
+    seconds * MILLIS_PER_SECOND * 1_000_000 + nanoseconds
+}
+
+impl RecordBatchIterator for ParquetFile {
+    fn schema(&self) -> &Arc<Schema> {
+        &self.projection_schema
+    }
+
+    fn next(&mut self) -> Result<Option<RecordBatch>> {
+        // advance the row group reader if necessary
+        if self.current_row_group.is_none() {
```

Comment:
at some point we may need to think about small row groups versus big batch size, so that we may need to read across row group boundaries.